kernel/main.c

```diff
@@ -340,6 +353,116 @@ async_init(void)
       panic("Unable to create a kthread for do_async_init()");
 }
 
+static void
+setup_uefi_runtime_services(void)
+{
+   struct mem_region ma;
+
+   if (!booting_with_uefi)
+      return;
+
+   EFI_RUNTIME_SERVICES *RT = TO_PTR(uefi_rt_addr);
+   printk("Runtime Services: %p\n", RT);
+
+   int num_entries = 0;
+
+   /*
+    * When SetVirtualAddressMap() is called, the physical
+    * addresses should be directly accessible. So, identity
+    * map them.
+    */
+   for (int i = 0; i < get_mem_regions_count(); i++) {
+      get_mem_region(i, &ma);
+
+      if (ma.type != TILCK_BOOT_EFI_RUNTIME_CODE && ma.type != TILCK_BOOT_EFI_RUNTIME_DATA)
+         continue;
+
+      const ulong pbegin = (ulong)ma.addr;
+      const ulong pend = MIN((ulong)(ma.addr+ma.len), (ulong)LINEAR_MAPPING_SIZE);
+
+      const size_t page_count = (pend - pbegin) >> PAGE_SHIFT;
+
+      ulong *vbegin = TO_PTR(pbegin);
+
+      /*
+       * SetupVirtualAddressMap() requires the runtime data
+       * as well as runtime code to be writable. So use
+       * `PAGING_FL_RW` for both.
+       */
+      size_t count =
+         map_pages(get_kernel_pdir(),
```

Comment:
OK, that works, but it believe it will be much better to have the first loop simply count the number of entries and nothing else. Map the pages in the next loop, before or assigning the descriptor. This loop could be simple as:
```
for (int i = 0; i < get_mem_regions_count(); i++) {

   if (ma.type == TILCK_BOOT_EFI_RUNTIME_CODE ||
       ma.type == TILCK_BOOT_EFI_RUNTIME_DATA)
   {
      num_entries++;
   }
}
```